### ISSUE TITLE
C++ migration

### ISSUE THREAD

**Issue by LecrisUT** [spglib/spglib#262]
Actually I've never asked. Are you guys comfortable with C++, and at what level? E.g.:
- Given a reimplementation, are you able to maintain it?
- Are you comfortable with class, smart pointer, and/or templating design? Would you want to learn through this process?

Thankfully it is possible to make gradual transitions for this, while keeping the C-api intact. In a nutshell, I would do it by making a C++ interface that calls the C interface, and then gradually make translations from one to the other in no particular order.

Main benefits that we can get:
- better memory management with less memory leaks
- better bindings since the intent and lifetime of the pointers can be passed to the other api
- thread safety
- simpler design through classes
- namespaces and simpler namings
- micro-optimization. Yes this is a negative, but the design exercise could be beneficial for other things down the line

**Comment by atztogo** [comment]
If gradual migration is possible, it can be our choice. But the effort is not negligible. So the guys have to be rewarded from this job. To think about this before starting is rather important.

**Comment by LecrisUT** [comment]
Indeed, that's why I wanted to open this issue to get everyone's position on this. I am biassed towards C++, but I've often seen some issues that are more easily solved there. At the very least we can create a C++ header file for better interface for that language.

**Comment by lan496** [comment]
At least, I am more comfortable with C++ than C although my knowledge almost stops at C++11. I agree with the benefits of migrating to C++ (or other more modern languages), but I am still overwhelmed by the non-trivial tasks because the current codes seem to have a lot of uncommented workarounds to search for symmetry in edge cases.

**Comment by LecrisUT** [comment]
My suggestion is to go top-down:
1. Create a C++ header with the design we want
2. Link the C++ class methods to the C functions. (this would be inefficient initially, but a smoother transition)
3. Add the unit-tests for the C++ side
4. Begin the migration
  - Copy the implementation from the C side with appropriate modern improvements
  - Create an `extern "C"` alias in the `.cpp` file that will be linked to the original `spglib.h`
  - Implement detailed unit-tests and documentation as we go
  - The unit-tests have to pass on both cpp and c interface side

We will steadily understand these workarounds as we go, and possibly there are more elegant ways around them.

> although my knowledge almost stops at C++11

That is actually perfect, that is the main standard that can bring the most advantages over C.

Also on my side, I am comfortable with C++20, and I can help on the `template`, `concept` and `const expr`. C++20 modules are still 2-3 years away from proper implementation.

**Comment by LecrisUT** [comment]
I want to add some notes here about C++ interface, it should have 3 types of inputs:
- `std::vector`, and `std::array` (initially not templatized, only native `int`, `double` and classes input). This should be the main interface
- `std::mdspan` with fallback to `std::span` + `static_cast`. This would be the main interface layer with C style arrays. In the backend, this is what will be used to call the current C interface
- `xtensor` interface. This seems like quite a strong tool that I think it will be beneficial to offer native (optional) support. Particularly I have high hopes for `xframe` which might allow to have a `pandas`/`xarray` like interface with different coordinates system

**Comment by atztogo** [comment]
@lan496 attempts migration in rust, https://github.com/spglib/moyo. I think it is a good choice, too. What do you think, @LecrisUT?

**Comment by LecrisUT** [comment]
I haven't done any rust, but I appreciate the coding practices that rust enforces. But the key question is if we can create bindings from `rust` to `python` and other languages. In C++ we have three ways:
- manually: wrap C++ classes with `class_init`, `class_end` kind of functions exposing raw pointers then using the normal bindings
  - disadvantage: it is prone to memory-leak
- manually: use built-in integrations to C++ like in pybind and other such methods
  - advantage: better memory management
  - disadvantage: we are limited to the languages that we can use for binding
- automated: use swig to automatically create bindings
  - advantage: wide range of languages
  - advantage: good memory management
  - advantage: integration with language standards like numpy

@lan496 What are your thoughts on rust -> C or rust -> C++ bindings. If you can do the latter I can help implement the c++ -> python, ruby, etc. languages via swig. What I've found so far is https://rust-lang.github.io/rust-bindgen/cpp.html, but I would need an example to see:
- Are smart pointers (`unique_ptr`, `shared_ptr`, etc.) and containers available in that binding?
- What is the build/linking process to C++/rust? E.g. how should a CMake project import from a rust project?

**Comment by lan496** [comment]
@LecrisUT Thanks for your consideration!
First of all, this Rust migration has just begun. Because I want to re-design algorithms for space-group search, it may take time (also this is a non-funded side project). I am considering rust -> C binding, and gradually replacing functions in spglib. From the code dependency, it will proceed as symmetry.h -> primitive.h -> spacegroup.h -> spglib.h (and other files).

I will use [Rust's FFI](https://doc.rust-lang.org/nomicon/ffi.html), but actually I am just learning it. Sorry, I cannot say the specific details like smart pointers for now.
